pkg/controller/nodemaintenance/nodemaintenance_controller.go

```diff
@@ -55,268 +133,487 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
+
+	case NodeStateCordon:
+		{
+			if drainer, kclient, err := r.initDrainer(); err == nil {
+				if err := r.runCordonOrUncordon(drainer, kclient, nodeName, true); err != nil {
+					// TODO: is that the right thing? or should it rather report error.
+					return reconcile.Result{Requeue: true, RequeueAfter: RequeuDrainingWaitTime * time.Second}, nil
+				}
+			} else {
+				return reconcile.Result{}, err
+			}
+			statusAnnoData.Status = NodeStateEvicting
+			_, err := r.setStatusAnnotation(node, statusAnnoData)
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+		fallthrough
+
+	case NodeStateEvicting:
+		{
+			if drainer, _, err := r.initDrainer(); err == nil {
+				if err := drain.RunNodeDrain(drainer, node.Name); err != nil {
+
+					// TODO: distinguish timeout errors from real errors where the reconcile failed.
+					r.reqLogger.Warningf("drain failed for machine %q: %v", node.Name, err)
+					return reconcile.Result{Requeue: true, RequeueAfter: RequeuDrainingWaitTime * time.Second}, nil
+				}
+			} else {
+				return reconcile.Result{}, err
+			}
+
+			statusAnnoData.Status = NodeStateMaintenanceMode
+			_, err := r.setStatusAnnotation(node, statusAnnoData)
+			if err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+		fallthrough
+
+	case NodeStateMaintenanceMode:
+		{
+			r.reqLogger.Info("maintenance mode")
 		}
-		// Error reading the object - requeue the request.
-		reqLogger.Info("Error reading the request object, requeuing.")
-		return reconcile.Result{}, err
 	}
+	return reconcile.Result{}, nil
+}
 
-	// Add finalizer when object is created
-	if instance.ObjectMeta.DeletionTimestamp.IsZero() {
-		if !ContainsString(instance.ObjectMeta.Finalizers, kubevirtv1alpha1.NodeMaintenanceFinalizer) {
-			instance.ObjectMeta.Finalizers = append(instance.ObjectMeta.Finalizers, kubevirtv1alpha1.NodeMaintenanceFinalizer)
-			if err := r.client.Update(context.TODO(), instance); err != nil {
-				return r.reconcileAndError(instance, err)
+func (r *ReconcileNodeMaintenance) processMaintModeOff(statusAnnoData *ReconcileNodeMaintenanceStatusInfo, node *corev1.Node) (reconcile.Result, error) {
+
+	nodeName := node.ObjectMeta.Name
+
+	switch statusAnnoData.Status {
+	case NodeStateEvicting:
+	case NodeStateMaintenanceMode:
+		{
+			statusAnnoData.Status = NodeStateUnEvicting
+			_, err := r.setStatusAnnotation(node, statusAnnoData)
+			if err != nil {
+				return reconcile.Result{}, err
 			}
 		}
-	} else {
-		// The object is being deleted
-		if ContainsString(instance.ObjectMeta.Finalizers, kubevirtv1alpha1.NodeMaintenanceFinalizer) {
-			// Stop node maintenance - uncordon and remove live migration taint from the node.
-			if err := r.stopNodeMaintenance(instance.Spec.NodeName); err != nil {
-				if errors.IsNotFound(err) == false {
-					return r.reconcileAndError(instance, err)
+		fallthrough
+
+	case NodeStateUnEvicting:
+		{
+			// cancelling evictions
+			if drainer, _, err := r.initDrainer(); err == nil {
```

Comment:
Similar to ``processMaintModeOn``, do things based on the state of the system not  ``statusAnnoData.Status``

For example, in this case, "am i about to loose the lease and are there pods being evicted?"   

This might imply that it is necessary to add some amount of extra time when creating the lease so that the controller has a window where it has the authority to undo the work of ``processMaintModeOn``

Reply:
done

Reply:
Did you consider the second part?